src/decoder/plugins/UsfDecoderPlugin.cxx

```diff
@@ -0,0 +1,342 @@
+
+static int stdio_fseek( void * f, int64_t offset, int whence )
+{
+    return fseek( (FILE*) f, offset, whence );
+}
+
+static int stdio_fclose( void * f )
+{
+    return fclose( (FILE*) f );
+}
+
+static long stdio_ftell( void * f )
+{
+    return ftell( (FILE*) f );
+}
+
+psf_file_callbacks stdio_callbacks =
+{
+    "\\/:",
+    stdio_fopen,
+    stdio_fread,
+    stdio_fseek,
+    stdio_fclose,
+    stdio_ftell
+};
+
+struct usf_loader_state
+{
+    uint32_t enable_compare;
+    uint32_t enable_fifo_full;
+    char * lib;
+    void * emu;
+};
+
+static int
+usf_loader(void * context, const uint8_t * exe, size_t exe_size,
+               const uint8_t * reserved, size_t reserved_size)
+{
+    struct usf_loader_state * state = ( struct usf_loader_state * ) context;
+
+    if ( exe && exe_size > 0 ) return -1;
+
+    return usf_upload_section( state->emu, reserved, reserved_size );
+}
+
+static int
+usf_info(void * context, const char * name, const char * value)
+{
+    struct usf_loader_state * state = ( struct usf_loader_state * ) context;
+
+    if ( strcasecmp( name, "_enablecompare" ) == 0 && strlen( value ) )
+        state->enable_compare = 1;
+    else if ( strcasecmp( name, "_enablefifofull" ) == 0 && strlen( value ) )
+        state->enable_fifo_full = 1;
+
+    return 0;
+}
+
+struct usf_tags {
+    std::string title;
+    std::string artist;
+    std::string composer;
+    std::string game;
+    std::string year;
+    std::string copyright;
+    std::string usfby;
+    std::string genre;
+    std::string length;
+    std::string fade;
+    std::string track;
+};
+
+static int
+usf_save_tag(const char * field, const char * name, const char *value, std::string& tag)
+{
+    if (strcasecmp(name, field) == 0) {
+        tag = value;
+        return 1;
+    }
+    return 0;
+}
+
+static int
+usf_tags_target(void * context, const char * name, const char * value)
+{
+    struct usf_tags * tags = ( struct usf_tags * ) context;
+    if (!value) return 0;
+
+    if (usf_save_tag("title", name, value, tags->title));
+    else if (usf_save_tag("artist", name, value, tags->artist));
+    else if (usf_save_tag("composer", name, value, tags->composer));
+    else if (usf_save_tag("game", name, value, tags->game));
+    else if (usf_save_tag("year", name, value, tags->year));
+    else if (usf_save_tag("copyright", name, value, tags->copyright));
+    else if (usf_save_tag("usfby", name, value, tags->usfby));
+    else if (usf_save_tag("genre", name, value, tags->genre));
+    else if (usf_save_tag("length", name, value, tags->length));
+    else if (usf_save_tag("fade", name, value, tags->fade));
+    else if (usf_save_tag("track", name, value, tags->track)){}
+    return 0;
+}
+
+static double
+get_length_from_string(std::string length)
+{
+    const char * str_length = length.c_str();
+    double ttime = 0, temp = 0, mult = 1, level = 1;
+    char Source[1024];
+    char * src = Source + strlen(str_length);
+    strcpy(&Source[1], str_length);
```

Comment:
Never never never ever use `strcpy()` unless you really really know what you're doing! This is an exploitable bug you just wrote.